File: MaximumProductOfThreeNumbers.cpp
#include <vector>
#include <cstring>
using namespace std;


// heap sort
class Solution {
public:
    int maximumProduct(vector<int>& nums) {
        vector<int>pos,neg;
        for(int n:nums){
            if(n<0){
                for(auto it=neg.begin();;it++){
                    if(it==neg.end()){neg.push_back(n);break;}
                    else if(*it>n){neg.insert(it,n);break;}
                }
                if(neg.size()>3)neg.pop_back();
            }

            if(n>=0){
                for(auto it=pos.begin();;it++){
                    if(it==pos.end()){pos.push_back(n);break;}
                    else if(*it<n){pos.insert(it,n);break;}
                }
                if(pos.size()>3)pos.pop_back();
            }
        }
        int result=INT_MIN;
        for(int i=0;i<neg.size();i++)pos.push_back(neg[i]);
        for(int i=0;i<pos.size();i++){
            for(int j=i+1;j<pos.size();j++){
                for(int k=j+1;k<pos.size();k++){
                    if(result<pos[i]*pos[j]*pos[k])result=pos[i]*pos[j]*pos[k];
                }
            }
        }
        return result;
    }
};

File: Maximum_Product_of_Three_Numbers.kt
import kotlin.math.max

class Solution {
    fun maximumProduct(nums: IntArray): Int {
        nums.sort()
        val n = nums.size
        if (nums.last() <= 0 || nums[0] >= 0) return nums[n - 1] * nums[n - 2] * nums[n - 3]
        return max(nums[n - 1] * nums[n - 2] * nums[n - 3], nums[0] * nums[1] * nums[n - 1])
    }
}

fun main() {
    val solution = Solution()
    val testCases = arrayOf(intArrayOf(1, 2, 3), intArrayOf(1, 2, 3, 4), intArrayOf(-1, -2, -3), intArrayOf(-1, -2, -3, 3))
    for (case in testCases) {
        println(solution.maximumProduct(case))
    }
}